docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,265 @@
+- from: 'src/**/*.js'
+  to: app/
+  flatten: true  # default to false
+  
+# Existing use-case: Copy all js under 'src' into 'dest', re-creating the directory structure below 'src'  
+# This corresponds to the current `'src/***/*.js': dest/`
+- from: 'src/**/*.js'
+  to: dest/
+  strip: 'src'   # default to ""
+  
+# New use-case: Copy all python files and infer their destination.
+- from: '**/*.py'
+  inferTo: true   # <- tbd, default to false
+```
+
+When determining the destination for a given file, all sync rules will be considered.
+Thus, a single file may be copied to multiple destinations.
+
+Error cases:
+- It should be an error to specify `inferTo=true` together with `to` in the same sync rule.
+  ```yaml
+  - from: '*.js'
+    to: dest/
+    inferTo: true
+  ```
+- It should be an error to specify either of `flatten` or `strip` together with `inferTo`.
+  ```yaml
+  - from: '*.js'
+    inferTo: true
+    flatten: true        # ERR
+  - from: 'src/**/*.js'
+    inferTo: true
+    strip: src/          # ERR
+  ```
+- It should be an error to specify `strip` with a prefix which which does not match the static prefix of the `from` field.
+  ```yaml
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: src/sub        # ERR
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: app/           # ERR
+  ```
+
+Open questions about the configuration:
+
+**\<Is the syntax clear enough?\>**
+Is the syntax clear?
+Is the configuration intuitive to use?
+Is there any ambiguity which needs attention?
+
+Resolution: __Not Yet Resolved__
+
+**\<Does it improve clarity to have the `inferTo` field?\>**
+A different approach could be to fall back to inference, if the `to` field is left blank.
+This would have the advantage that there needs to be no validation about having `to` and `inferTo` in the same rule, which should cause an error.
+
+Resolution: __Not Yet Resolved__
+   
+**\<Error for `strip` and `flatten`\>**
+Given the new sync config, users may try to combine various flags in unintended ways.
+For example, should `strip` and `flatten` in the same rule always be an error?
+My intuition would say, that `strip+flatten` is the same as `flatten`, but this should to be discussed.
+
+Resolution: __Not Yet Resolved__
+   
+   
+#### Migration plan
+The new configuration supports all existing use-cases and is fully backwards compatible.
+An automatic schema upgrade shall be implemented.
+
+### Open Issues/Question
+
+**\<Overlapping copies\>** 
+Overlapping copies can be caused by at least two ways:
+
+1. The `COPY/ADD` instructions in the Dockerfile may copy to the same destination. For example:
+   ```dockerfile
+   COPY foo /foo
+   COPY bar/ /   # bar contains a file called "foo"
+   ```
+2. Two sync rules may copy files to the same destination. For example:
+   ```yaml
+   sync:
+   - from: foo
+     to: /baz
+   - from: bar
+     to: /baz
+   ```
+   
+Should such cases be treated as errors, warnings, or not be detected at all?
```

Comment:
We can `warn` for now and then `error` later.
Right now, these do not get detected.
Changing it to `warn` will give users time to fix their config and still be able to proceed with their normal workflow.
After that, we can change this to `error`

WDYT?

Reply:
So for case 1 where the overlapping copies come from the builder, we should follow what the builders do. Docker does not treat this as warnings, so neither should we. However, our syncing must be consistent with the builders. In this case, we need to follow the "last-one-wins" policy.

For case 2, where the overlapping copies may come from (mis-)configuration, I am not a big fan of warnings, too. This is for two reasons:
1. The unix copy command also does not warn about existing destination paths. Instead it simply overwrites. The sync rules are simply a list of automated copy instructions, so the same safety guard should apply (namely none).
2. Overlapping destination paths are costly to detect. Yet, this check needs to run for every sync just to cover an edge case.

Reply:
Makes sense. I think when you run skaffold dev with `--verbose` we do see all the computed sync files and users can use that to debug.